enumerable.rb

```diff
@@ -0,0 +1,109 @@
+module Enumerable
+  def my_each
+    # Should return the array except block is given
+    return to_enum unless block_given?
+
+    each do |i|
+      yield i
+    end
+  end
+
+  def my_each_with_index
+    # Should return the array except block is given
+    return to_enum unless block_given?
+
+    (0..size - 1).my_each do |i|
+      yield(self[i], i)
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    # an empty array to store our selected element
+    arr = []
+    (0...size - 1).my_each do |i|
+      arr << self[i] if yield(self[i])
+    end
+    arr
+  end
+
+  def my_all?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 unless yield self[i]
+    end
+    count.zero?
+  end
+
+  def my_any?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 if yield self[i]
+    end
+    count != 0
+  end
+
+  def my_none?
+    status = true
+    my_each do |x|
+      status = false if x.nil? || x == false
+    end
+    return status unless block_given?
+
+    count = 0
+    size.times do |i|
+      count += 1 if yield(self[i])
+    end
+    count.zero?
+  end
+
+  def my_count(variable = nil)
+    count = 0
+    (0..size - 1).my_each do |i|
+      if !block_given? && variable.nil?
+        count = size
+      elsif block_given?
+        count += 1 if yield(self[i])
+      else
+        count += 1 unless self[i] != variable
+      end
+    end
+    count
+  end
+
+  def my_map(&proc)
+    arr = []
+    return to_enum unless block_given?
+
+    (0..size - 1).my_each do |i|
+      puts
+      arr << yield(self[i]) if !proc && yield(self[i])
+      arr << yield(self[i]) if proc&.call(self[i])
+    end
+    arr
+  end
+
+  def my_inject(variable = 0)
+    (0..size - 1).my_each do |i|
+      variable = yield(variable, self[i])
+    end
+    variable
+  end
```

Comment:
#### my_inject 

- [ ] combines all elements of enum by applying a binary operation, specified by a block

  > (1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.

> [2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13